File: backend/app/routers/keys.py
import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from backend.app.database import get_db
from backend.app.dependencies import get_current_user
from backend.app.models.user import User
from backend.app.schemas.key import PublicKeyOut, PublishKeyRequest
from backend.app.services import key_service

router = APIRouter(prefix="/keys", tags=["keys"])


@router.post("/publish", response_model=PublicKeyOut)
async def publish(
    body: PublishKeyRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await key_service.publish_key(db, current_user.id, body.public_key_b64)


@router.get("/{user_id}", response_model=PublicKeyOut)
async def get_key(
    user_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await key_service.get_active_key(db, user_id)


@router.get("/batch/", response_model=list[PublicKeyOut])
async def get_batch(
    user_ids: str = Query(...),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    ids = [uuid.UUID(uid.strip()) for uid in user_ids.split(",") if uid.strip()]
    return await key_service.get_batch_keys(db, ids)

File: backend/routes/keys.js
const express = require('express');
const auth = require('../middleware/auth');
const keyService = require('../services/keyService');
const router = express.Router();

// Publish public encryption key
router.post('/publish', auth, async (req, res, next) => {
  try {
    const { publicKeyB64 } = req.body;
    const key = await keyService.publishKey(req.user.id, publicKeyB64);
    res.status(201).json(key);
  } catch (err) {
    next(err);
  }
});

// Get user's active public key
router.get('/:userId', auth, async (req, res, next) => {
  try {
    const key = await keyService.getKey(req.params.userId, req.user.id);
    res.json(key);
  } catch (err) {
    next(err);
  }
});

// Batch fetch public keys
router.get('/batch', auth, async (req, res, next) => {
  try {
    const { user_ids } = req.query;
    if (!user_ids) throw new Error('user_ids query parameter required');
    const ids = user_ids.split(',');
    const keys = await keyService.batchGetKeys(ids, req.user.id);
    res.json(keys);
  } catch (err) {
    next(err);
  }
});

module.exports = router;
